File: tools/hollow-knight-patches/src/skins/core/SkinCoreModels.cs
// C# 9 records require this compiler marker on netstandard2.1, not a runtime dependency.
#if NETSTANDARD2_1
namespace System.Runtime.CompilerServices { internal static class IsExternalInit { } }
#endif

namespace DualSouls.Skins.HollowKnight.Core
{
    public sealed record DeathEpoch(ulong Value);
    // Trusted caller correlates one before/after pair with a fresh monotonically increasing value.
    public sealed record DeathOccurrenceToken(ulong Value);

    // Hero identity; unchanged across deaths while the hero identity remains the same.
    public sealed record HeroBindingToken(string Value);
    public sealed record SkinBindingToken(string Value);
    public sealed record StableRespawnToken(DeathEpoch DeathEpoch, HeroBindingToken Hero, SkinBindingToken Skin);

    public sealed record HeroObservation(
        HeroBindingToken Hero, SkinBindingToken Skin, bool AcceptingInput,
        bool FullDamageMode, int Health, bool CanTakeDamage, bool Playable,
        bool Paused, bool Cutscene, bool SceneTransition);

    // Zero means no event yet. Both high-water values survive delivery and rebind.
    public sealed record LifecycleState(
        HeroBindingToken ArmedHero = null,
        DeathEpoch PendingEpoch = null,
        int StableCount = 0,
        HeroBindingToken CurrentHero = null,
        SkinBindingToken CurrentSkin = null,
        DeathEpoch LastConfirmedEpoch = null,
        SkinBindingToken ArmedSkin = null,
        DeathOccurrenceToken ArmedOccurrence = null,
        DeathOccurrenceToken OccurrenceHighWater = null)
    {
        public DeathOccurrenceToken OccurrenceHighWater { get; init; } = OccurrenceHighWater ?? new DeathOccurrenceToken(0);
        public DeathEpoch LastConfirmedEpoch { get; init; } = LastConfirmedEpoch ?? new DeathEpoch(0);
    }

    public abstract record LifecycleSignal
    {
        private LifecycleSignal() { }
        // Trusted caller boundary; updates cannot appoint binding authority.
        public sealed record Rebind(HeroBindingToken Hero, SkinBindingToken Skin) : LifecycleSignal;
        public sealed record BeforeDeath(HeroBindingToken Hero, SkinBindingToken Skin, DeathOccurrenceToken Occurrence) : LifecycleSignal;
        public sealed record AfterDeath(HeroBindingToken Hero, SkinBindingToken Skin, DeathOccurrenceToken Occurrence) : LifecycleSignal;
        public sealed record Update(HeroObservation Observation) : LifecycleSignal;
    }

    public sealed record LifecycleDecision(
        LifecycleState State, DeathEpoch ConfirmedEpoch, StableRespawnToken StableToken, string Diagnosis);
}

File: src/SilksongLauncher.Launcher/app/src/main/kotlin/dev/silksong/launcher/skins/core/SkinCoreModels.kt
package dev.silksong.launcher.skins.core

import dev.silksong.launcher.skins.registry.SkinBindingToken

@JvmInline
value class DeathEpoch(val value: ULong)

/** Trusted caller correlates one before/after pair with a fresh monotonically increasing value. */
@JvmInline
value class DeathOccurrenceToken(val value: ULong)

/** Hero identity; unchanged across deaths while the hero identity remains the same. */
@JvmInline
value class HeroBindingToken(val value: String)

data class StableRespawnToken(val deathEpoch: DeathEpoch, val hero: HeroBindingToken, val skin: SkinBindingToken)

data class HeroObservation(
    val hero: HeroBindingToken,
    val skin: SkinBindingToken,
    val acceptingInput: Boolean,
    val fullDamageMode: Boolean,
    val health: Int,
    val canTakeDamage: Boolean,
    val playable: Boolean,
    val paused: Boolean,
    val cutscene: Boolean,
    val sceneTransition: Boolean,
)

/** Zero means no event yet. Both high-water values survive delivery and rebind. */
data class LifecycleState(
    val armedHero: HeroBindingToken? = null,
    val pendingEpoch: DeathEpoch? = null,
    val stableCount: Int = 0,
    val currentHero: HeroBindingToken? = null,
    val currentSkin: SkinBindingToken? = null,
    val lastConfirmedEpoch: DeathEpoch = DeathEpoch(0uL),
    val armedSkin: SkinBindingToken? = null,
    val armedOccurrence: DeathOccurrenceToken? = null,
    val occurrenceHighWater: DeathOccurrenceToken = DeathOccurrenceToken(0uL),
)

sealed interface LifecycleSignal {
    /** Trusted caller boundary; updates cannot appoint binding authority. */
    data class Rebind(val hero: HeroBindingToken, val skin: SkinBindingToken) : LifecycleSignal
    data class BeforeDeath(val hero: HeroBindingToken, val skin: SkinBindingToken, val occurrence: DeathOccurrenceToken) : LifecycleSignal
    data class AfterDeath(val hero: HeroBindingToken, val skin: SkinBindingToken, val occurrence: DeathOccurrenceToken) : LifecycleSignal
    data class Update(val observation: HeroObservation) : LifecycleSignal
}

data class LifecycleDecision(
    val state: LifecycleState,
    val confirmedEpoch: DeathEpoch?,
    val stableToken: StableRespawnToken?,
    val diagnosis: String,
)
